### PR TITLE
Fix the compilation error that occurs when opening avx series instructions

### PR DESCRIPTION
**Pull request by monchickey**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

An error will occur when building ClickHouse with the avx series of instructions enabled.
CMake command：
```shell
cmake .. -DCMAKE_BUILD_TYPE=Release -DENABLE_AVX=ON -DENABLE_AVX2=ON -DENABLE_AVX2_FOR_SPEC_OP=ON
```
Failed message：
```
[1558/11630] Building CXX object contrib/snappy-cmake/CMakeFiles/_snappy.dir/__/snappy/snappy.cc.o
FAILED: contrib/snappy-cmake/CMakeFiles/_snappy.dir/__/snappy/snappy.cc.o 
/usr/bin/ccache /usr/bin/clang++-17 --target=x86_64-linux-gnu --sysroot=/opt/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc -DHAVE_CONFIG_H -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -I/opt/ClickHouse/base/glibc-compatibility/memcpy -isystem /opt/ClickHouse/contrib/snappy -isystem /opt/ClickHouse/build/contrib/snappy-cmake -isystem /opt/ClickHouse/contrib/llvm-project/libcxx/include -isystem /opt/ClickHouse/contrib/llvm-project/libcxxabi/include -isystem /opt/ClickHouse/contrib/libunwind/include --gcc-toolchain=/opt/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64 -fdiagnostics-color=always -Wno-enum-constexpr-conversion -fsized-deallocation  -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -mavx -mavx2 -fasynchronous-unwind-tables -ffile-prefix-map=/opt/ClickHouse=. -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -w -O3 -DNDEBUG   -D OS_LINUX -Werror -nostdinc++ -std=c++2b -MD -MT contrib/snappy-cmake/CMakeFiles/_snappy.dir/__/snappy/snappy.cc.o -MF contrib/snappy-cmake/CMakeFiles/_snappy.dir/__/snappy/snappy.cc.o.d -o contrib/snappy-cmake/CMakeFiles/_snappy.dir/__/snappy/snappy.cc.o -c /opt/ClickHouse/contrib/snappy/snappy.cc
/opt/ClickHouse/contrib/snappy/snappy.cc:1061:3: error: unknown type name '__m256i'
 1061 |   __m256i data = _mm256_lddqu_si256(static_cast<const __m256i *>(src));
      |   ^
/opt/ClickHouse/contrib/snappy/snappy.cc:1061:55: error: unknown type name '__m256i'
 1061 |   __m256i data = _mm256_lddqu_si256(static_cast<const __m256i *>(src));
      |                                                       ^
/opt/ClickHouse/contrib/snappy/snappy.cc:1061:18: error: use of undeclared identifier '_mm256_lddqu_si256'; did you mean '_mm_lddqu_si128'?
 1061 |   __m256i data = _mm256_lddqu_si256(static_cast<const __m256i *>(src));
      |                  ^~~~~~~~~~~~~~~~~~
      |                  _mm_lddqu_si128
/usr/lib/llvm-17/lib/clang/17/include/pmmintrin.h:38:1: note: '_mm_lddqu_si128' declared here
   38 | _mm_lddqu_si128(__m128i_u const *__p)
      | ^
/opt/ClickHouse/contrib/snappy/snappy.cc:1061:66: error: cannot initialize a parameter of type 'const __m128i_u *' with an lvalue of type 'const void *'
 1061 |   __m256i data = _mm256_lddqu_si256(static_cast<const __m256i *>(src));
      |                                                                  ^~~
/usr/lib/llvm-17/lib/clang/17/include/pmmintrin.h:38:34: note: passing argument to parameter '__p' here
   38 | _mm_lddqu_si128(__m128i_u const *__p)
      |                                  ^
/opt/ClickHouse/contrib/snappy/snappy.cc:1062:40: error: unknown type name '__m256i'
 1062 |   _mm256_storeu_si256(reinterpret_cast<__m256i *>(dst), data);
      |                                        ^
/opt/ClickHouse/contrib/snappy/snappy.cc:1065:49: error: unknown type name '__m256i'
 1065 |     data = _mm256_lddqu_si256(static_cast<const __m256i *>(src) + 1);
      |                                                 ^
/opt/ClickHouse/contrib/snappy/snappy.cc:1065:65: error: arithmetic on a pointer to void
 1065 |     data = _mm256_lddqu_si256(static_cast<const __m256i *>(src) + 1);
      |                                                            ~~~  ^
/opt/ClickHouse/contrib/snappy/snappy.cc:1066:42: error: unknown type name '__m256i'
 1066 |     _mm256_storeu_si256(reinterpret_cast<__m256i *>(dst) + 1, data);
      |                                          ^
8 errors generated.
[1567/11630] Building CXX object contrib/rocksdb-cma...rocksdb.dir/__/rocksdb/db/arena_wrapped_db_iter.cc.o
ninja: build stopped: subcommand failed.
```
The reason is that snappy does not enable SNAPPY_HAVE_X86_CRC32.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit a1ced8c88f4a28b9e2a5112139f7941165ead9f0 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55049/a1ced8c88f4a28b9e2a5112139f7941165ead9f0/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>❌ failure</td></tr>
<tbody>
</table>

File: contrib/snappy-cmake/CMakeLists.txt [hunk 1]
```diff
@@ -24,6 +24,12 @@ else ()
     set (SNAPPY_HAVE_SSSE3 0)
 endif ()
 
+if (ARCH_AMD64 AND (ENABLE_AVX OR ENABLE_AVX2 OR ENABLE_AVX512))
```
Comment (nickitat) [inline contrib/snappy-cmake/CMakeLists.txt:27]: `crc32` is part of `sse4.2`, so we can relax the condition

Comment (monchickey) [reply contrib/snappy-cmake/CMakeLists.txt:27]: Can it be changed to the following?
```cmake
if (ARCH_AMD64 AND (ENABLE_SSSE3 OR ENABLE_AVX OR ENABLE_AVX2 OR ENABLE_AVX512))
    set (SNAPPY_HAVE_SSSE3 1)
    set (SNAPPY_HAVE_X86_CRC32 1)
else ()
    set (SNAPPY_HAVE_SSSE3 0)
    set (SNAPPY_HAVE_X86_CRC32 0)
endif ()
```
Thanks.

Comment (nickitat) [reply contrib/snappy-cmake/CMakeLists.txt:27]: currently we have a pretty strange situation. we already set the flag for `ssse3`

https://github.com/ClickHouse/ClickHouse/blob/master/contrib/snappy-cmake/CMakeLists.txt#L21-L25

so I think the condition should be `if (ARCH_AMD64 AND ENABLE_SSE42)`

Comment (monchickey) [reply contrib/snappy-cmake/CMakeLists.txt:27]: Yes, I tried it and it works.
```cmake
if (ARCH_AMD64 AND ENABLE_SSSE3)
    set (SNAPPY_HAVE_SSSE3 1)
else ()
    set (SNAPPY_HAVE_SSSE3 0)
endif ()

if (ARCH_AMD64 AND ENABLE_SSE42)
    set (SNAPPY_HAVE_X86_CRC32 1)
else ()
    set (SNAPPY_HAVE_X86_CRC32 0)
endif ()
```
How about it?

Comment (nickitat) [reply contrib/snappy-cmake/CMakeLists.txt:27]: looks good

**Review by nickitat — APPROVED** [review]

**Pull request merged**